crates/ide-assists/src/handlers/extract_module.rs

```diff
@@ -380,7 +380,25 @@ impl Module {
         }
 
         for (vis, syntax) in replacements {
-            add_change_vis(vis, syntax.first_child_or_token());
+            let item = syntax.children_with_tokens().find(|node_or_token| {
+                match node_or_token.kind() {
+                    // We're looking for the start of functions, impls, structs, traits, and other documentable/attribute
+                    // macroable items that would have pub(crate) in front of it
+                    SyntaxKind::FN_KW
+                    | SyntaxKind::STRUCT_KW
+                    | SyntaxKind::ENUM_KW
+                    | SyntaxKind::TRAIT_KW
+                    | SyntaxKind::TYPE_KW
+                    | SyntaxKind::CONST_KW
+                    | SyntaxKind::MOD_KW => true,
+                    // If we didn't find a keyword, we want to cover the record fields in a struct
+                    SyntaxKind::NAME => true,
+                    // Otherwise, the token shouldn't have pub(crate) before it
+                    _ => false,
+                }
+            });
+
+            add_change_vis(vis, item);
```

Comment:
This logic seems inverted to me, we should pick the first token that is NOT a comment or attribute, so we should skip `SyntaxKind::COMMENT`, `SyntaxKind::ATTR` and (I think also) `SyntaxKind::WHITESPACE` and pick the first thing that is not one of those 3. That way we don't have to explicitly list the things that we expect should be there at the start.

Reply:
I agree, realized a few cases this already didn't cover, such as `async`, `extern`, and `union`. Explicitly skipping these three kinds works for the tests in extract_module.rs.